Importing Molecules from PQR

etp12 opened the pull request:
This extension allows users to search for molecules from the Pitt Quantum Repository, preview the molecule, and then download it and open it in Avogadro. It is accessed under File > Import From PQR in the menu.

Screenshot of the UI and molecule preview: ![](https://i.gyazo.com/e52cce26da392cace106a7d6315a0696.png)

After clicking "download": ![](https://i.gyazo.com/29a5faa0e87e638471596b647b0230ac.png)

cryos commented on avogadro/qtplugins/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 include_directories(SYSTEM ${EIGEN3_INCLUDE_DIR})
 
-find_package(Qt5 COMPONENTS Widgets OpenGL Network Concurrent REQUIRED)
+find_package(Qt5 COMPONENTS Widgets OpenGL Network Concurrent WebView REQUIRED)
```
I don't think this is needed, and the others should be removed. There was a bug I was working around, we can leave this in for now, and I will fix all of them.

cryos commented on avogadro/qtplugins/importpqr/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+find_package(Qt5WebEngineWidgets REQUIRED)
```
Please use find_package(Qt5 COMPONENTS WebEngineWidgets REQUIRED) as this only needs Qt5_DIR to be set, finding all other modules relative to that.

cryos commented on avogadro/qtplugins/importpqr/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+
+# Extension
+set(importpqr_srcs
+  ImportPQR.cpp
```
Please just use lower case for all source, header, ui, etc files.

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.cpp:
```diff
@@ -0,0 +1,264 @@
+*/
+void PQRRequest::sendRequest(QString url, QString mol2, QString downloadFolder)
+{
+	QUrl httpRequest(url);
```
Please use 2 spaces for indentation

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.cpp:
```diff
@@ -0,0 +1,264 @@
+  if(results == NULL) {
+    return QString("N/A");
+  }
+		QString mol2 = results[num].mol2url;
```
This looks over indented

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.cpp:
```diff
@@ -0,0 +1,264 @@
+{
+  std::string str = formula.toStdString();
+  QString toReturn;
+  for(int i = 0; i < str.length(); i++) {
```
Prefer space after for and if keywords.

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.cpp:
```diff
@@ -0,0 +1,264 @@
+* @param formula The formula string
+*/
+float PQRRequest::getMolMass(QString formula) {
+	std::string str = formula.toStdString();
```
This might be tab vs spaces, but this has varying indentation

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.cpp:
```diff
@@ -0,0 +1,264 @@
+			if (islower(str[i + 1])) {
+        element = { str[i], str[i + 1] };
+				//this might be the last element of the formula
+        if(isdigit(str[i + 2])) {
```
Missing space after if

cryos commented on avogadro/qtplugins/importpqr/PQRRequest.h:
```diff
@@ -0,0 +1,146 @@
+
+class PQRRequest : public QObject
+{
+	Q_OBJECT
```
Indentation

cryos commented:
It looks nice, there are a bunch of indentation issues, maybe tabs and spaces intermixed? We use lowercase file names, this looks like a great addition, didn't have chance to give it a try but will later or tomorrow.

cryos commented:
I had a chance to give this a try on Linux. The form does not resize at all, so if I make the dialog bigger it doesn't use any of the extra space (and it just adds scroll bars when smaller). It is suggesting a download folder of C:/mol2 which won't work on Linux (or Mac I would suspect). I was able to search by name, but the preview remains blank. I would also expect this to go inside the new import item as a submenu item. Nothing happened when I clicked Download, even after adding a valid path, but I would expect to be able to just download the file as a string and open without ever saving just as download by name does.

I haven't looked into the cause of any of the issues, but wanted to provide some usage feedback

cryos commented:
I think it works more than I first realized, when I double clicked on the entry then the preview worked, and then with a valid path it was able to download and display the molecule (I guess the download button should only be enabled when it sees a molecule as selected, and it should be marked as selected when the row is selected).

etp12 commented:
I can get those formatting issues and UI issues fixed ASAP. As for the download folder, should it just be removed completely and automatically save the file to a temporary location to be deleted later after the molecule is loaded into Avogadro?

etp12 commented:
I originally misread your comment about the download folder, it now just reads the response string just like NetworkDatabases. I changed the indentation issues, moved the extension to the right submenu, and only enable the download button when a result is selected. 

I'll comment again when I fix the resizing issue.

etp12 commented:
I altered the ui a little bit, but Qt's layouts seem kind of tricky to work with. Let me know if this version seems a little better.

ghutchis commented:
@cryos I'm happy to fix any remaining layout / UI issues once merged.

cryos commented:
I was hoping to take a look at this earlier, trying to compile it on my laptop it turns out that Ubuntu doesn't seem to have packaged the Qt5 WebView module - I have been looking for a package for a while, and Google is confused with all the webkit stuff. This looks like a newer component on the desktop, it would be nice to optionally disable the dependency, we can default to on if you like but I would like to be able to continue developing on my laptop, and don't particularly need the plugin to do that.

Ethan seems to have a bad set up on most of the commits where his name is not properly attributed - do you mind me fixing the branch up with proper author name, and making the webview dependency optional?

cryos commented:
#136 fixes the authorship on the commits, and makes the webview module optional.

ghutchis commented:
I don't think we need to have QWebView per se.. certainly having a preview widget is highly useful. QWebEngineView might be an alternative. We initially tried using Qt's native SVG preview and it did worse than pretending the SVG was part of a webpage.

cryos commented:
Yeah, I think web view is likely the best option, packaging webkit is a pretty heavy dependency for a preview, hopefully webview is relatively easy to package, I don't think web engine would be easy. PR #136 fixes up authorship, and I think can be merged, we can take a look at packaging. It seems like a reasonable middle ground for now, happy to iterate on this more after merging. PNGs would be easiest, but I doubt you want to generate them.

ghutchis commented:
We could probably do SVG -> PNG for the server, but I think there will be other use cases where rendering web objects will be useful

cryos commented:
This looks much better, there are still some minor tweaks. I am going to close this PR, and merge the other PR that simply fixes authorship and makes the plugin optionally build depending on whether we could find the Qt5 WebView module.
